compiler.py

```diff
@@ -0,0 +1,637 @@
+from .constants import *          # definitions of various constants mapping Python entitites to mlog equivalents
+
+
+@dataclass
+class Function:
+    """This stores information about a user-defined function that is callable within a compilable program."""
+    start: int   #  Line number for the start of the function within the compiled program
+    argc: int    #  Count of number of arguments the function takes
+
+
+class CompilerError(ValueError):
+    def __init__(self, code, node: ast.AST):
+        super().__init__(f'[{code}/{node.lineno}:{node.col_offset}] {ERROR_DESCRIPTIONS[code]}')
+
+
+class Compiler(ast.NodeVisitor):
+    """ast.NodeVisitor is a superclass with methods that enable straightforward visitation of the various nodes
+    in an AST syntax tree, calling corresponding methods for the different types of nodes as they're visited, where
+    parent nodes typically opt to initiate visits for their children.  This subclass overwrites the defauly visit
+    methods with ones that generate compiled mlog code as the various parts of the tree are visited."""
+    def __init__(self):
+        self._ins = [
+            f'set {REG_STACK} 0',
+        ] # This will end up being a list of instructions that will compose the compiled code.
+          # We begin by initializing the stack
+        #TODO I'm not sure we really want/need a stack by default -- probly better to do everything with dummy variables
+        self._in_def = None   # will be set to a True-ish value while visiting user-defined function definitions
+        self._functions = {}  # will map user defined function names to Function objects containing info about functions
+
+    def compile(self, code):
+        """This is the main function for the compiler taking in a chunk of Python code (specified either as a
+        string, *** a function whose source code file can be found via python's inspect module, or *** a Path
+        indicating a source code file to compile.  This visits various nodes in the syntax tree of that code,
+        amassing a list of mlog instructions that will perform the equivalent of that code in mindustry.
+        This list is then concatenated into a single string, and returned."""
+        # TODO enable this to handle function and Path inputs, not just strings
+        self.visit(ast.parse(code))  # visit the root node of the syntax tree, which will then visit branches and leaves
+        return self.generate_masm()  # concatenate our amassed list of instructions into a single string to return
+        # TODO print the resulting mlog instructions to console, so user can read or copy/paste
+        # TODO place a copy of the resulting mlog string in the clipboard, so it can be pasted to Mindustry
+
+    def visit(self, node):
+        """This will call some specific flavor of visit method, depending upon what type of node this is, typically
+        one of the ones defined below."""
+        super().visit(node)

```

Comment:
Not our job to document other modules. I also realized this definition is redundant and can be deleted (in another pull request), I forgot to do so after I removed some debugging code.